### ISSUE TITLE
[ENHANCEMENT] Migrating jails between Bastille servers

### ISSUE THREAD

**Issue by VulcanRidr** [BastilleBSD/bastille#679]
I am in the process of migrating off of iocage, bastille is awesome so far. One of the very cool features is that you can "hot" export jails. 

I have two mini-servers (Lenovo Thinkcentre m710qs) that are my jail servers. A really nice feature  would be for some type of telemetry between the bastille processes to talk to each other, and be able to hot swap jails "on the fly".  

Initially, it would be nice to extend bastille export functionality you could do something like "bastille export --host <remote hostname> <jail>...Then through the telemetry link, shut down the jail on the old host.

I wrote a set of scripts for iocage where I could iocage export and then copy the files over to the other host, and shut the jail down on the first host that I would be happy to share.

**Comment by cedrictineo** [comment]
@VulcanRidr I'd be thankful to get a look at your iocage migration scripts if you'd accept to share.

About Bastille migration of jails between hosts, I stumbled upon this article, let us know if that works for you and solves it. But yes, I upvote that - in-the-tool migration of jails between hosts would be great, even for the purpose of getting a host ready to take over in case of failure of the first one.

https://hackacad.net/post/2023-04-28-migrating-multiple-bastille-jails-new-host/

Slightly out of topic: I'm looking for info/ressources to be more HA with jails - if you have any info, articles, I'm very interested. Would be great to do HA jails with Bastille.

**Comment by VulcanRidr** [comment]
Sure! You may get them from https://github.com/VulcanRidr/iocage_jail_mgmt. Be warned they work, but they are really ugly and kludgey. 

I will take a look at that hackacad article. 

Now that I have things more fully stood up, I can give you some details on my setup, and other thoughts on how I am using it...I have two jail hosts, erebus and equinox, running Bastille. I have mounted the backups directory on both to my TrueNAS box, so all backups are shared. 

I am converting all of my jails from thin to thick jails, since they are long-term servers, and since I am having problems upgrading jails from 13.2 to 14.0. So I usually run all of the jails on one host or the other, but especially while I am converting the jails, I have all of the 13.2 ones on equinox, and then recreating them on 14.0 on erebus. But I can see other reasons why I would need to split jails across hosts. 

So these are some of my thoughts. I know you can use the bastille_list directive to tell bastille which jails to start, but what about 
* An option in bastille to associate one or more bastille servers, and then set flags for each jail servers as to which jails are "primary" on that jail.
* A check whenever starting a jail to see if it is already running elsewhere.
* a way to do something like "bastille move <jail> <server>" 
* You could also stop a jail on one server and start it manually (or create it) on a different one, as I am doing with my jails to convert to thick and 14.0.

**Comment by VulcanRidr** [comment]
You know, as I was mulling this over, another thought occurred to me. Since, unlike iocage, bastille takes advantage of thesnapshot feature of ZFS, instead of going down the route of exporting and importing jails, you could easily migrate jails by doing a zfs send /usr/bastille/<jail> and any associated datasets (root, pgsql, mysql, etc). I always create separate pgsql and mysql datasets, since the ideal recordsizes are 8k and 16k respectively, though it appears I am going to have to fix my database datasets for the jails since they are set at 1M...Sigh. 

In any case, I believe a mechanism that allows you to migrate jails through the zfs send | zfs recv would be faster than exporting and importing...

**Comment by cedrictineo** [comment]
Thanks for the info you provide - this puts me on the trail to try to setup bastille (i want to move from Iocage to Bastille) and to set-up the migration or quick switch-over (in case of the main host going offline) of bastille-managed-jails from one host to another - through zfs send/rcv...

**Comment by VulcanRidr** [comment]
Cedric, just for the record, I have changed the way that I have my backups set up. I have started NFS mounting my /usr/local/bastille/backups from my NAS box on both  jail servers. Then I run something similar to the following to back up the jails:

bastille list | grep -v Hostname | awk '{ print $3 }' | while read x ; do bastille export --xz $x ; done

which makes the jail backups immediately available on both servers.

**Comment by VulcanRidr** [comment]
Cedric, just for the record, I have changed the way that I have my backups set up. I have started NFS mounting my /usr/local/bastille/backups from my NAS box on both  jail servers. Then I run something similar to the following to back up the jails:

bastille list | grep -v Hostname | awk '{ print $3 }' | while read x ; do bastille export --xz $x ; done

which makes the jail backups immediately available on both servers. I could write a wrapper script around this, but am considering the possibility/feasability of doing jail management through ansible.

**Comment by bmac2** [comment]
@VulcanRidr  is there an issue to do on this ticket?  Seems you found a solution that works with your script.

**Comment by VulcanRidr** [comment]
@bmac2, what I was looking for was something where you could instrument bastille to be able to "push" jails between two bastille hosts, instead of backing up on one host, then when complete, going to the other host, turning it off (and possibly deleting it) before you turn on the jail on the new host...I already share /usr/local/bastille/backups between all bastille hosts (I technically have 3 since I have one on my desktop for spinning up boxes before they go onto the servers).

Instead, I was thinking, as it is a ZFS dataset, is it safe to zfs send | zfs recv  /usr/local/bastille/jails/<jailname> from one host to the other, or does more need to be migrated?

**Comment by 2Belette** [comment]
@VulcanRidr I have two new servers ready and waiting for 14.2 next week to test Bastille with theses two servers and would like to achieve exactly what you are describing ! I also think that leveraging zfs send/recv would be a nice solution but as you described Bastille should have some mechanism to be aware to automatically put the jails in some states.
This would make Bastille to another huge step if we can accomplish that.

**Comment by bmac2** [comment]
https://bastille.readthedocs.io/en/latest/chapters/migration.html  is the instructions to move from iocage to bastille, FYI.

**Comment by tschettervictor** [comment]
You can use https://www.bolthole.com/zrep to sync two boxes with failover included.

**Comment by tschettervictor** [comment]
@VulcanRidr @SnapshotCiTy @2Belette 

There is #1019

New migrate command.

Feedback welcome

**Comment by bmac2** [comment]
OK, code for this is in master branch.  @2Belette @VulcanRidr @SnapshotCiTy   if you guys can pull the master code and test that would  be great.  This migrates a jail between servers and starts it up on its new home and IF you want use -d switch to delete the old jail.   STILL BETA and EXPERIMENTAL, but it worked for @tschettervictor and I both.  Please test and report any bugs you find with it.  Closing this issue as addressed.

**Comment by 2Belette** [comment]
First of all, many thanks for this initiative, I think this was one of the last big rock missing to Bastille!

Some preliminaries comments from me (I have just started to test the command not yet until the end) : 
- we should not rely on ping to check if the host is up/down, instead we should use nc on port 22 (I have example where icmp can be block between vlans but not ssh and Bastille thinks the host is down)
- we should support both sudo and doas (my servers don't have sudo installed, only doas)
- we should give the option to set HOST like HOST:PORT when ssh is not on port 22 (not my case but thinking about others users)

**Comment by 2Belette** [comment]
Regarding the command itself, I got some issues (most of the command is OK, just the very end of -x is showing issue) : 

```
[ERROR]: Archive 'jail-c3_2025-05-07-223632.xz' not found.
+ migrate_cleanup jail-c3 blt 10.7.30.1
+ local '_jail=jail-c3'
+ local '_user=blt'
+ local '_host=10.7.30.1'
+ ssh -i /.ssh/id_rsa blt@10.7.30.1 sudo rm -f '/jail-c3_*.*'
zsh:1: no matches found: /jail-c3_*.*
+ rm -f /usr/local/bastille/migrate/jail-c3_2025-05-07-223632.sha256 /usr/local/bastille/migrate/jail-c3_2025-05-07-223632.xz
+ error_exit '[ERROR]: Failed to import jail on remote system.'
+ error_notify '[ERROR]: Failed to import jail on remote system.'
+ echo -e '\033[0;31m[ERROR]: Failed to import jail on remote system.\033[0;0m'
[ERROR]: Failed to import jail on remote system.
```

**Comment by tschettervictor** [comment]
Good ideas there. Will get them implemented.

Can you confirm that your config file on the remote system has the "bastille_migratedir" variable?

**Comment by tschettervictor** [comment]
@2Belette 

Syntax is not `bastille migrate TARGET USER@HOST:PORT` with the PORT being optional. Defaults to 22.

You can now also set `--doas` to use `doas` as the su command.

nc is now used to check host uptime

**Comment by 2Belette** [comment]
> Good ideas there. Will get them implemented.
> 
> Can you confirm that your config file on the remote system has the "bastille_migratedir" variable?
> 
> This is the same error I had when that variable was not present. The value should be auto added using "bastille setup --filesystem" when you run the migrate command.
> 
> Do you also mind sharing a full -x output?
> 
> EDIT: Your remote system must also be updated to latest code...

they are both using latest git pull

bastille_migratedir="${bastille_prefix}/migrate" is also on both systems bastille.conf

**Comment by tschettervictor** [comment]
The reason you are seeing the above issue is because it didn't fill the "_migrate_user" with `doas -u‘

How could we ID the user with doas?

Is it even necessary? We could simply use the user that is supplied by the user.

**Comment by 2Belette** [comment]
Yes I agree this should be enough to use the user supplied

**Comment by tschettervictor** [comment]
Can you test #1032?

We now just use USER as the user instead of doing -u whoami, which is redundant anyway.

**Comment by 2Belette** [comment]
Thanks comments added to the PR

**Comment by 2Belette** [comment]
Going back to this one, perhaps a very small enhancement, I now that the format is `bastille migrate [option(s)] TARGET USER@HOST[:PORT] `but I think we should make the user optional, if not used we should try the current user, like ssh is doing.

I am trying : `doas bastille migrate -l --doas client1 blt@10.10.10.1 ` which is working as expected but if I try : `doas bastille migrate -l --doas client1 10.10.10.1` I would expect it to use my current user, I got instead this error : `[ERROR]: Could not find keys for user: 10.10.10.1`

Not a big deal but a small "nice to have"

**Comment by tschettervictor** [comment]
That's possible. Will see what we can do.

**Comment by tschettervictor** [comment]
Test 1169
@2Belette

**Comment by 2Belette** [comment]
Thanks, it is not working for me as it is trying to use root instead of my user.
The reason is because I am running the command with doas, but this should be the same issue with sudo.

In parallel, I found another issue, migrate.sh is only trying to find id_rsa, so it is not working with other key types. On my side I have a id_ed25519, but for others it can be also another file, I think it should look at id_* instead
